src/ExplainProvenanceRecords.h

```diff
@@ -0,0 +1,271 @@
+                }
+            }
+        }
+    }
+
+    RamDomain getLabel(std::string relName, std::vector<RamDomain> e) {
+        auto key = std::make_pair(relName, e);
+
+        if (valuesToLabel.find(key) != valuesToLabel.end()) {
+            return valuesToLabel[key];
+        } else {
+            return -1;
+        }
+    }
+
+    std::vector<RamDomain> getTuple(std::string relName, RamDomain l) {
+        auto key = std::make_pair(relName, l);
+
+        if (labelToValue.find(key) != labelToValue.end()) {
+            return labelToValue[key];
+        } else {
+            return std::vector<RamDomain>();
+        }
+    }
+
+    std::vector<RamDomain> getSubproofs(std::string relName, RamDomain l) {
+        auto key = std::make_pair(relName, l);
+
+        if (labelToProof.find(key) != labelToProof.end()) {
+            return labelToProof[key];
+        } else {
+            return std::vector<RamDomain>();
+        }
+    }
+
+    std::vector<std::string> getInfo(std::string relName) {
+        assert(info.find(relName) != info.end());
+        return info[relName];
+    }
+
+    std::string getRule(std::string relName, int ruleNum) {
+        auto key = std::make_pair(relName, ruleNum);
+
+        if (rule.find(key) != rule.end()) {
+            return rule[key];
+        } else {
+            return std::string("no rule found");
+        }
+    }
+};
+
+class ExplainProvenanceRecords : public ExplainProvenance {
+private:
+    ProvenanceInfo provInfo;
+    std::unique_ptr<TreeNode> root;
+
+public:
+    ExplainProvenanceRecords(SouffleProgram& p) : ExplainProvenance(p), provInfo(p) {
+        setup();
+    }
+
+    void setup() override {
+        provInfo.setup();
+    }
+
+    std::unique_ptr<TreeNode> explainSubproof(std::string relName, RamDomain label, size_t depth) override {
+        bool isEDB = true;
+        bool found = false;
+        for (auto rel : prog.getAllRelations()) {
+            if (rel->getName().find(relName) != std::string::npos) {
+                found = true;
+            }
+            std::regex provRelName(relName + "-provenance-[0-9]+");
+            if (std::regex_match(rel->getName(), provRelName)) {
+                isEDB = false;
+                break;
+            }
+        }
+
+        // check that relation is in the program
+        if (!found) {
+            return std::unique_ptr<TreeNode>(new LeafNode("Relation " + relName + " not found"));
+        }
+
+        // if EDB relation, make a leaf node in the tree
+        if (prog.getRelation(relName) != nullptr && isEDB) {
+            auto subproof = provInfo.getTuple(relName + "-output", label);
+
+            // construct label
+            std::stringstream tup;
+            tup << join(numsToArgs(relName, subproof), ", ");
+            std::string lab = relName + "(" + tup.str() + ")";
+
+            // leaf node
+            return std::unique_ptr<TreeNode>(new LeafNode(lab));
```

Comment:
I think you can use `return std::make_unique<LeafNode>(lab)`.
The result is the same, but you can avoid using new, thus avoid reflexively thinking about whether you need to delete.